sdb/target.py

```diff
@@ -0,0 +1,87 @@
+#
+# Copyright 2019 Delphix
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""
+This module attempts to abstract out most of the functionality
+provided by the drgn.Program that we are using for the current
+context of a running SDB instance. Its current form can be
+thought of as a singleton class where the infrastructure will
+initialize the `prog` global once and the consumers of the
+infrastructure will be using the helper methods defined in
+it to query information from the target. That said, there are
+still edge cases when consumers may still want the actual
+drgn.Program object (e.g. they are using the stack_trace() or
+list_for_each() helpers of the drgn API) and since global
+variables cannot explicitly imported from other modules without
+the star (*) notation, we provide the get_prog() wrapper.
```

Comment:
`prog` is defined in `target.py` but we want to expose it to consumers through the `sdb` package (e.g. in `sdb/__init__.py` which is imported every time a new file does `import sdb`).

In other words we would like new commands to just use `import sdb` instead of
```
import sdb
import sdb.target
```
In order to better avoid import cycles in the future and provide a simpler API.

Now if we do the following in `sdb/__init__.py`
```
from sdb.target import prog
```
That `prog` is not re-exported from that module, so consumers cannot just access `prog` by doing `import sdb` or `from sdb import prog`. There are a couple of ways of re-exporting a variable like this through `__init__.py`:
```
# [1] The most common that is mentioned in the comment but also the ugliest/hackiest one because it imports everything
from sdb.target import *
# [2] The "redundant" one - I can switch the code to that if people feel it is better
from sdb.target import prog as prog
# [3] The most explicit one (reads as "this module's namespace consists of prog" - I can switch the code to that too if people feel it is better
from sdb.target import prog
__all__ = ['prog']
```
After thinking hard about the above though, I realized that maybe if I just expose an API of `sdb.target` to `sdb` that can get re-exposed it would lead to what I think is a more modular solution.

Does this answer your question? Also, can you think of a better way to write the comment that you've highlighted above?

Reply:
Yea, so I guess the take away is, we don't want folks to use `sdb.target` directly?

e.g. while they could do this (which is what I was expecting they would/could do, to get access to `prog`):
```
from sdb.target import prog
```
in the specific command.. we'd rather them not, and just rely on:
```
import sdb
```
as the "public interface".. does that sound right? where, they'd then get access to the `get_prog` function..

I think adding some commentary about the fact that we don't want consumers to use `sdb.target` directly, would help explain the situation.

Reply:
yeah that's correct. ok I will change the comment accordingly